### PR TITLE
Add rule mutation implementation, parallel training, and OpenEvolve baseline

### PR DESCRIPTION
**Pull request by IKACE**
- Add mutation training mode (`train-evolution`), parallel LLM training (`train-LLM-only-parallel`), and the new MutateAgent. Adds full-metric evaluation (`eval_metrics/base.py`, `event_f1pa.py`, `point_f1pa.py`, `metrics.py`), decision-tree and OpenEvolve baselines, a benchmark harness (`runtime/benchmark.py`), and rule-mutation test assets.
- Updates `driver.py` / `runtime/engine.py` with `--p_cores`, `--rule_per_group`, and `--max_iter` flags; propagates validation-set evaluation through rule selection so the TrainPerfSelector can rank rules by event-F1 on the val split.
- Fixes a few bugs uncovered while testing: `sklearn.metrics.precision_recall_curve` kwarg rename (`probas_pred` → `y_score`); train/val eval results were written as flattened dicts while the selector expected full dicts; `run_iteration_thread` referenced `eval_res` before assignment when the review loop timed out.
- Updates `README.md` — removes the Aggregator (`train-combined-*`) documentation and replaces it with descriptions and examples for the three supported modes (`train-LLM-only`, `train-LLM-only-parallel`, `train-evolution`). 
- Bumps new transitive deps (`tiktoken`) while keeping the security-pinned versions of `certifi`, `h11`, `requests`, `urllib3`.

### CODE REVIEW

**Comment by IKACE** [comment]
@microsoft-github-policy-service agree

File: baseline/openevolve/evaluator.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    Only supports evaluation helpers (eval + eval_scores_by_metrics).
+    """
+
+    def __init__(self, chunk_size: int, mode: str = "train-LLM-only") -> None:
+        self.LLM = None
+        self.name = "ReviewAgentEvalOnly"
+        self.dataset = None
+        self.dataset_name = None
+        self.chunk_size = chunk_size
+        self.mode = "train-LLM-only"
```
[the remark this answers is not in the document]
Comment (IKACE) [reply baseline/openevolve/evaluator.py:34]: This is intended as the mode for OpenEvolve evaluator should be independent of what is used in Argos.

File: baseline/openevolve/initial_program.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    # your comment to describe how normal data behave
+    # Normal Rule 1
+    # Normal Rule 2
+    # your code to detect if the given sample is abnormal
+    # Abnormal Rule 1
+    if ...
+    # Abnormal Rule 2
+    if ...
+    # return labels as a 1d numpy array indicating abnormal/normal of each index
+
```
[the remark this answers is not in the document]
Comment (IKACE) [reply baseline/openevolve/initial_program.py:13]: The initial program in OpenEvolve is intended to be the same as what is used in Argos.

File: agent/test/rule_test.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import numpy as np
```
Comment (abuccts) [inline agent/test/rule_test.py]: what's the usage for files under `test/`? are they unit tests or something else?

File: agent/test/rule_test.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    
+    # Abnormal Rule 3: Check if the last 4 values deviate more than 20% from the mean
+    last_four = sample[-4:, 0]
+    lower_bound = mean * 0.8
+    upper_bound = mean * 1.2
+
+    if np.all(last_four < lower_bound) or np.all(last_four > upper_bound):
+        labels[-1] = 1
+
+    return labels
```
Comment (abuccts) [inline agent/test/rule_test.py]: add a newline ending in every file, you can use [.editorconfig](https://editorconfig.org/)

File: agent/agent.py [hunk 5]
```diff
@@ -7,21 +7,24 @@
...
+TIMEOUT = 60 * 60
 TIMEOUT_FIRST_REVIEW = 2 * 60
-TIMEOUT_LLM = 5 * 60
-TIMEOUT_PER_REVIEW = 2 * 60
+TIMEOUT_LLM = 20 * 60 * 10
+TIMEOUT_PER_REVIEW = 20 * 60 * 10
 TIMEOUT_IMAGE = 10 * 60
 TIMEOUT_INFERENCE = 1 * 60
 
+SELF_HOSTED_LLM_LIST = ["meta-llama/Llama-3.1-8B-Instruct", "deepseek-ai/DeepSeek-R1-Distill-Qwen-14B"]
```
Comment (abuccts) [inline agent/agent.py]: create a config yaml for these configurations? you can use [omegaconf](https://github.com/omry/omegaconf)

File: agent/mutate_agent.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+probaility_mutate_rules={
+    "remove(s)": 0.30, # remove rule (otherwise llm just keep adding rule) 0.05
+    "free": 0.30, # free to mutate -> default                              0.84
+    "same": 0.0, # keep the same                                          0.8
+    "change-only": 0.30, # change the rule                                0.99
+    "remove-all-but-one": 0.10 # remove all but one rule                  1.0  
+}
+
+def generate_mutate_prompt(chunk_size, type="free"):
+    start_prompt = f"""You are a computer science researcher working at DeepMind. You will be given a Python function that contains rules to detech cloud anomaly in microservices.
```
Comment (abuccts) [inline agent/mutate_agent.py]: move the prompts to a separate file?

File: agent/mutate_agent.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+- add noise to sample from a distribution
+"""
+
+# probaility_mutate_rules={
+#     "remove(s)": 0.05, # remove rule (otherwise llm just keep adding rule) 0.05
+#     "free": 0.79, # free to mutate -> default                              0.84
+#     "same": 0.0, # keep the same                                          0.8
+#     "change-only": 0.15, # change the rule                                0.99
+#     "remove-all-but-one": 0.01 # remove all but one rule                  1.0  
+# }
```
Comment (abuccts) [inline agent/mutate_agent.py]: remove?

**Review by abuccts — COMMENTED** [review]
please resolve the merge conflict as well

File: agent/test/rule_test.py (hunk 3 above)
Comment (IKACE) [reply agent/test/rule_test.py]: these are legacy tests for testing mutation, removed

File: agent/test/rule_test.py (hunk 4 above)
Comment (IKACE) [reply agent/test/rule_test.py]: updated

File: agent/agent.py (hunk 5 above)
Comment (IKACE) [reply agent/agent.py]: updated, moved to a unified config file under `config/agent.yaml`

File: agent/mutate_agent.py (hunk 6 above)
Comment (IKACE) [reply agent/mutate_agent.py]: updated, move all agent prompts to `agent/prompts`

File: agent/mutate_agent.py (hunk 7 above)
Comment (IKACE) [reply agent/mutate_agent.py]: removed

**Review by abuccts — APPROVED** [review]
